wandb/sdk/launch/sweeps/scheduler_sweep.py

```diff
@@ -120,40 +113,53 @@ def _heartbeat(self, worker_id: int) -> None:
+        return True
 
     def _run(self) -> None:
         # Go through all workers and heartbeat
-        for worker_id in self._workers.keys():
-            self._heartbeat(worker_id)
-        try:
-            run: SweepRun = self._heartbeat_queue.get(
-                timeout=self._heartbeat_queue_timeout
-            )
-        except queue.Empty:
-            wandb.termlog(f"{LOG_PREFIX}No jobs in Sweeps RunQueue, waiting...")
-            time.sleep(self._heartbeat_queue_sleep)
-            return
-        # If run is already stopped just ignore the request
-        if run.state in [
-            SimpleRunState.DEAD,
-            SimpleRunState.UNKNOWN,
-        ]:
-            return
-        wandb.termlog(
-            f"{LOG_PREFIX}Converting Sweep Run (RunID:{run.id}) to Launch Job"
-        )
-        _ = self._add_to_launch_queue(
-            run_id=run.id,
-            entry_point=["python3", run.program] if run.program else None,
-            # Use legacy sweep utilities to extract args dict from agent heartbeat run.args
-            config={
-                "overrides": {
-                    "run_config": LegacySweepAgent._create_command_args(
-                        {"args": run.args}
-                    )["args_dict"]
-                }
-            },
-        )
+        for worker_id in self._workers:
+            if not self._heartbeat(worker_id):
+                break
+
+        for _worker_id in self._workers:
+            try:
+                run: SweepRun = self._heartbeat_queue.get(
+                    timeout=self._heartbeat_queue_timeout
+                )
+
+                # If run is already stopped just ignore the request
+                if run.state in [
+                    SimpleRunState.DEAD,
+                    SimpleRunState.UNKNOWN,
+                ]:
+                    if self.state == SchedulerState.FINISHING:
+                        # Scheduler hit the run_cap, let it launch runs then it will stop
+                        pass
+                    else:
+                        logging.debug(
+                            f"{LOG_PREFIX}Can't launch run: {run.id} in state {run.state}"
+                        )
+                        continue
+
+                wandb.termlog(
+                    f"{LOG_PREFIX}Converting Sweep Run (RunID:{run.id}) to Launch Job"
+                )
+                _ = self._add_to_launch_queue(
+                    run_id=run.id,
+                    entry_point=["python3", run.program] if run.program else None,
+                    # Use legacy sweep utilities to extract args dict from agent heartbeat run.args
+                    config={
+                        "overrides": {
+                            "run_config": LegacySweepAgent._create_command_args(
+                                {"args": run.args}
+                            )["args_dict"]
+                        }
+                    },
+                )
+            except queue.Empty:
```

Comment:
Seems like somewhere if in the `FINISHING` state and the heartbeat queue is empty (no more runs to make) we should enter a loop polling on existing runs and run the `self.exit()` method when they are all done

Reply:
This is actually a bigger philosophical question right, like do you think the scheduler's job is to monitor all the runs or just schedule/add to launch queues? I was kinda thinking of it more like the latter, but obviously there could be reasons to do the former... @KyleGoyette

Reply:
The main reason I suggested polling was so that the exit hook doesn't terminate them prematurely. We just need to be sure that the exit method doesn't get other important aspects inserted into the method, that would not be called when the sweep hits its run cap

Reply:
Implemented polling, extendable to other conditions like "pause" if we pipe that in later. @KyleGoyette